utils/common/shared_pool_allocator.h

```diff
@@ -0,0 +1,195 @@
+
+This program is free software; you can redistribute it and/or
+modify it under the terms of the GNU General Public License
+as published by the Free Software Foundation; version 2 of
+the License.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program; if not, write to the Free Software
+Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
+MA 02110-1301, USA. */
+
+
+/* Makes PoolAllocator STL-compliant */
+
+#pragma once
+
+#include <stdlib.h>
+#include <algorithm>
+#include <boost/weak_ptr.hpp>
+#include <boost/shared_ptr.hpp>
+#include <boost/atomic.hpp>
+#include <boost/thread/thread.hpp>
+#include <boost/thread/mutex.hpp>
+#include <boost/thread/lock_guard.hpp>
+
+
+namespace utils {
+
+
+  struct iAllocMemUse {
+    virtual size_t getMemUsage() const = 0;
+  };
+
+  namespace _ {
+
+    template <uint32_t _val> struct next_pow2 {
+      static const uint32_t _val0 = _val - 1;
+      static const uint32_t _val1 =  _val0 | (_val0 >> 1);
+      static const uint32_t _val2 = _val1 | (_val1 >> 2);
+      static const uint32_t _val3 = _val2 | (_val2 >> 4);
+      static const uint32_t _val4 = _val3 | (_val3 >> 8);
+      static const uint32_t _val5 = _val4 | (_val4 >> 16);
+      static const uint32_t value = 1 + _val5;
+    };
+
+    template <size_t _len>
+    union alloc_node
+    {
+      alloc_node * _next;
+      uint8_t _buff[_len];
+    };
+
+    
+    template <size_t _bufflen>
+      struct alloc_stack : iAllocMemUse {
+      typedef boost::shared_ptr<alloc_stack> ptr;
+      typedef alloc_node <_bufflen> node_type;
+      static const size_t alloc_node_size = (sizeof(void*) > _bufflen ? sizeof(void*) : _bufflen);
+
+        /** returns a shared_ptr<alloc_stack> but maintains a weak_ptr<alloc_stack>
+         ** so consecutive callers will recieve the same instance until all callers are finished
+         ***/
+      static ptr get(size_t prealloc=0) {
+        static boost::weak_ptr<alloc_stack> oStack;
+        static boost::mutex mux;
+        ptr oRet;
+        bool bCreated = false;
+        {
+          boost::lock_guard<boost::mutex> oLock(mux);
+          oRet = oStack.lock();
+          if (!oRet) {
+            oRet = ptr(new alloc_stack(prealloc));
+            oStack = oRet;
+            bCreated = true;
+          }
+        }       
+        if (bCreated && prealloc) {
+          boost::thread(prepopulate).detach();
+        }
+        return oRet;
+
+      }
+
+      void * pop(){
+        for (;;) {                  
+          node_type * pRet = _root.load();
+          if (!pRet) {
+            void * pVoid = NULL;
+            if (posix_memalign(&pVoid, sizeof(void*), alloc_node_size) || !pVoid) throw std::bad_alloc();
```

Comment:
Hm, the destruction time as well.  Since these are all individually allocated nodes, they'd also have to be deallocated one by one when the alloc_stack goes away.